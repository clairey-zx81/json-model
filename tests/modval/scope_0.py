#! /bin/env python
#
# Model: ./scope_0.model.json
#
# Generated by jmc (JSON Model Compiler) (https://github.com/clairey-zx81/json-model)
#

from typing import Callable
import re as re
import datetime
import urllib.parse

type Jsonable = None|bool|int|float|str|list[Jsonable]|dict[str, Jsonable]
type Path = list[str]
type Report = list[str]|None
type CheckFun = Callable[[Jsonable, str, Report], bool]
type PropMap = dict[str, CheckFun]
type TagMap = dict[None|bool|float|int|str, CheckFun]

# extract type name
def _tname(value: Jsonable) -> str:
    return type(value).__name__

# maybe add message to report
def _rep(msg: str, rep: Report) -> bool:
    rep is None or rep.append(msg)
    return False

# regex "/[0-9]/"
_jm_re_0 = re.compile("[0-9]").search

# define "$S" ($.'$S')
def json_model_2(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $.'$S'
    # "/[0-9]/"
    result = isinstance(value, str) and _jm_re_0(value) is not None or _rep(f"does not match FESC at {path}", rep)
    if not result:
        rep is None or rep.append(f"not an expected REGEX at {path} [$.'$S']")
    return result

# define "$" ($)
def json_model_1(value: Jsonable, path: str, rep: Report = None) -> bool:
    # $
    result = isinstance(value, list) and len(value) == 2
    if result:
        # $.0
        result = json_model_2(value[0], path, rep)
        if not result:
            rep is None or rep.append(f"not an expected $S at {path + '.0'} [$.0]")
        if result:
            # $.1
            result = json_model_2(value[1], path, rep)
            if not result:
                rep is None or rep.append(f"not an expected $S at {path + '.1'} [$.1]")
    else:
        rep is None or rep.append(f"not array or bad array length at {path} [$]")
    if not result:
        rep is None or rep.append(f"not array or unexpected array at {path} [$]")
    return result

# entry function check_model
def check_model(value: Jsonable, path: str = "$", rep: Report = None) -> bool:
    return json_model_1(value, path, rep)


# possibly run as a script: $0 values...
if __name__ == "__main__":
    import json
    import sys
    for fn in sys.argv[1:]:
        try:
            with open(fn) as f:
                value = json.load(f)
            reasons = []
            if check_model(value, "", reasons):
                print(f"{fn}: PASS")
            else:
                print(f"{fn}: FAIL {reasons}")
        except Exception as e:
            print(f"{fn}: ERROR ({e})")

