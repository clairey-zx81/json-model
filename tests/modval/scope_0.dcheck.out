scope_0.0.true.json: True
scope_0.1.false.json: False [('', '$', 'bad array length')]
